lib_client/run_time.ml

```diff
@@ -6,244 +6,255 @@ let src =
+    | Queued_for of float (* elapsed time in seconds *)
+    | Running of { queued_for : float; ran_for : float }
+    | Finished of { queued_for : float; ran_for : float option }
+  [@@deriving show]
+
+  let queued_for = function
+    | Cached -> 0.
+    | Queued_for v -> v
+    | Running v -> v.queued_for
+    | Finished v -> v.queued_for
+
+  let ran_for = function
+    | Cached -> 0.
+    | Queued_for _ -> 0.
+    | Running v -> v.ran_for
+    | Finished v -> Option.value ~default:0. v.ran_for
+
+  let of_timestamp ~build_created_at ?(current_time = Unix.gettimeofday ())
+      (ts : Timestamp.t) : t =
+    match ts with
+    | Queued v -> Queued_for (current_time -. v)
+    | Running { queued_at; started_at } ->
+        Running
+          {
+            queued_for = started_at -. queued_at;
+            ran_for = current_time -. started_at;
+          }
+    | Finished { queued_at; started_at; finished_at } -> (
+        if finished_at < build_created_at then Cached
+        else
+          match started_at with
+          | None ->
+              Finished { queued_for = finished_at -. queued_at; ran_for = None }
+          | Some v ->
+              Finished
+                {
+                  queued_for = v -. queued_at;
+                  ran_for = Some (finished_at -. v);
+                })
+
+  let total = function
+    | Cached -> 0.
+    | Queued_for v -> v
+    | Running v -> v.queued_for +. v.ran_for
+    | Finished v -> v.queued_for +. Option.value ~default:0. v.ran_for
+end
+
+module Job = struct
+  let partition_build_steps build =
+    let analysis (ji : Client.job_info) =
+      let re = Str.regexp_string "analysis" in
+      try
+        ignore (Str.search_forward re ji.variant 0);
+        true
+      with Not_found -> false
+    in
+    let analysis_steps, rest = List.partition analysis build in
+    match analysis_steps with
+    | [] -> Error "No analysis step found"
+    | [ h ] -> Ok (h, rest)
+    | _ :: _ -> Error "Multiple analysis steps found"
+
+  let build_created_at ~build =
+    let analysis (ji : Client.job_info) =
+      let re = Str.regexp_string "analysis" in
+      try
+        ignore (Str.search_forward re ji.variant 0);
+        true
+      with Not_found -> false
+    in
+    let filtered = List.filter analysis build in
+    match filtered with
+    | [] -> Error "No analysis step found"
+    | [ h ] -> Ok h.queued_at
+    | _ :: _ -> Error "Multiple analysis steps found"
+
+  let total_of_run_times (build : Client.job_info list) =
+    let build_created_at =
+      Option.value ~default:0.
+        (Option.join @@ Result.to_option @@ build_created_at ~build)
     in
-    Ok (List.fold_left minn max_float jil)
+    build
+    |> List.filter_map (fun ji -> Result.to_option @@ Timestamp.of_job_info ji)
+    |> List.map (TimeInfo.of_timestamp ~build_created_at)
+    |> List.fold_left (fun subtotal rt -> subtotal +. TimeInfo.total rt) 0.
```

Comment:
This could be combined without to have to walk the list three times, but perhaps it doesn't matter too much.